winpr/libwinpr/sspi/Negotiate/negotiate.c

```diff
@@ -72,45 +99,184 @@ const SecPkgInfoW NEGOTIATE_SecPkgInfoW = {
 	NEGOTIATE_SecPkgInfoW_Comment /* Comment */
 };
 
-static void negotiate_SetSubPackage(NEGOTIATE_CONTEXT* context, const TCHAR* name)
+static const sspi_gss_OID_desc spnego_OID = { 6, "\x2b\x06\x01\x05\x05\x02" };
+static const sspi_gss_OID_desc kerberos_OID = { 9, "\x2a\x86\x48\x86\xf7\x12\x01\x02\x02" };
+static const sspi_gss_OID_desc kerberos_wrong_OID = { 9, "\x2a\x86\x48\x82\xf7\x12\x01\x02\x02" };
+static const sspi_gss_OID_desc ntlm_OID = { 10, "\x2b\x06\x01\x04\x01\x82\x37\x02\x02\x0a" };
+
+static const SecPkg SecPkgTable[] = {
+	{ KERBEROS_SSP_NAME, &KERBEROS_SecurityFunctionTableA, &KERBEROS_SecurityFunctionTableW },
+	{ NTLM_SSP_NAME, &NTLM_SecurityFunctionTableA, &NTLM_SecurityFunctionTableW }
+};
+
+static const Mech MechTable[] = {
+	{ &kerberos_OID, &SecPkgTable[0], 0, 0 },
+	{ &ntlm_OID, &SecPkgTable[1], 0, 0 },
+};
+
+static const int MECH_COUNT = sizeof(MechTable) / sizeof(Mech);
+
+enum NegState
 {
-	WINPR_ASSERT(context);
-	WINPR_ASSERT(name);
+	NOSTATE = -1,
+	ACCEPT_COMPLETED,
+	ACCEPT_INCOMPLETE,
+	REJECT,
+	REQUEST_MIC
+};
+
+typedef struct
+{
+	enum NegState negState;
+	BOOL init;
+	sspi_gss_OID_desc supportedMech;
+	SecBuffer mechTypes;
+	SecBuffer mechToken;
+	SecBuffer mic;
+} NegToken;
+
+static BYTE* negotiate_write_tlv(BYTE* buf, BYTE tag, ULONG len, BYTE* value)
```

Comment:
`value` could be `const BYTE*`
We don't use `ULONG` (`len` parameter) often for that kind of stuff `size_t` would be better